wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/LanguageServerConfigInitializer.java

```diff
@@ -110,6 +113,11 @@ void initialize() {
         communicationProviderRegistry.add(id, communicationProvider);
         localityRegistry.add(id, isLocal);
 
+        Map<String, String> attributes = config.getAttributes();
+        if (attributes.containsKey(PROJECTS_ROOT_ATTR_NAME)) {
+          projectsRootRegistry.add(id, attributes.get(PROJECTS_ROOT_ATTR_NAME));
+        }
+
```

Comment:
Would it not be better to have it as dedicated field (not as an attribute), IMO it is pretty reasonable to expect it for every LS?

Reply:
As it is potentially just an optional attribute I decided to add an attributes map to be more flexible on attributes in future. To me it seems more reasonable to have dedicated fields for mandatory attributes and an attribute map for all optional ones. In this context `RegexProvider#getFileWatchPatterns` probably should be moved to optional category as well, but it's a separate topic.

Reply:
Right, it is optional as for workspace configuration but I'd suppose it to always have some value (i.e. default) on lower level, no?

Reply:
Yes, I think I see what you mean. The way it is implemented right now it does not have to always have a value. I will consider a possibility of adding a dedicated field for this case, but I also would prefer to leave an map for optional attributes.

Reply:
Yes, I mean it would probably be good to encapsulate it in there. Otherwise, as long as QA passed, I am ok to improve it later